src/server/page_server.cpp

```diff
@@ -188,9 +188,33 @@ page_server::connection_handler::receive_stop_log_prior_dispatch (tran_server_co
 void
 page_server::connection_handler::receive_oldest_active_mvccid (tran_server_conn_t::sequenced_payload &a_sp)
 {
-  /* TODO dummy function. will be modified corretly. soon  */
-  const MVCCID oldest_mvccid = *reinterpret_cast<const MVCCID *const> (a_sp.pull_payload().c_str());
-  er_log_debug (ARG_FILE_LINE, "receive_oldest_active_mvccid(): %d\n", oldest_mvccid);
+  assert (m_server_type == transaction_server_type::PASSIVE);
+
+  const auto oldest_mvccid = *reinterpret_cast<const MVCCID *const> (a_sp.pull_payload().c_str());
+  const auto channel_id = get_channel_id ();
+  std::lock_guard<std::mutex> lockg { m_ps.m_pts_oldest_active_mvccids_mtx };
+
+  /*
+   * 1. The entry is already created when ths PTS is connected.
+   * 2. It is updated by the PTS only when it move foward.
+   *    Without update, it is MVCCID_ALL_VISIBLE by default, which is lesser than any mvccid assigned.
+   */
+  assert (m_ps.m_pts_oldest_active_mvccids.find (channel_id) != m_ps.m_pts_oldest_active_mvccids.end());
+  assert (m_ps.m_pts_oldest_active_mvccids[channel_id] < oldest_mvccid);
+
+  m_ps.m_pts_oldest_active_mvccids[channel_id] = oldest_mvccid;
```

Comment:
Can we use a simpler implementation - such as:
- in each `connection_handler` add a `m_pts_oldest_active_mvccids` whose access is protected by a mutex (it can even be an atomic value if that's more to your linking)
- initialize the value with `MVCCID_ALL_VISIBLE` in the ctor (this initialization does not need to be protected by the mutex as, in `set_passive_tran_server_connection`, the connection handler is first constructed and then added to the vector of connection handlers)
- `receive_oldest_active_mvccid` receives the value and stores it in the connection_handler member (thread safe)
- when the handler for the ATS request to retrieve the aggregated value, the vector for PTS connection handlers is iterated to aggregate the value.

Do we lose anything with this?

I think this approach solves the following:
- no need to have unique channel id's (I agree that those id's are not unique - but it's just a feature that was not needed until now)
- no need to make sure that the entry for that PTS is removed when the PTS either pro-actively disconnects from PS or upon abnormal disconnect to make sure the m_pts_oldest_active_mvccids stays in the correct shape
- no need to deal with active transaction server's connection handler when aggregating the final value as that one is not part of the vector
- a more decoupled approach as only the access to one's PTS's value is locked at one time (by the connection handler when receiving a value; and by the ATS's connection handler when retrieving the values to construct the aggregated value)
- the performance of the aggregator to calculate the value to be sent to ATS is unlikely to be a problem (but a pro-active calculation can be implemented if needed)

Reply:
I didn't consider it thoroughly, but I thought that it seems awkward for the "connection handler" to manage the oldest mvccid of the connected transaction server. I thought it should be in charge of just connecting and communicating. And, the cost for page_server to manage those values would not that expensive. To address it, we can make a wrapper class including the connection handler and the pts information, but I reckon it's too much for this. 
 Anyway, I agree your approach is way simpler. Just my concern. What do you think?

Reply:
You do have a point of polluting connection handler with data.
And a point about the extra class/structure.

A paranthesis:
I think we can make the channel id unique by introducing the `m_socket` into it. With the restriction that the function must not only be called after connect/accept, and after the socket number is ~= `INVALID_SOCKET`. Which I think is the case since `set_passive_tran_server_connection` is called after the call to `server_channel::accept` in `css_process_server_server_connect`.

Thus we can leave the bulk implementation the way you already implemented it.

Or you can add an extra struct/class to handle everything separately - with thread safe access from the side of the PTS connection handlers (update values) and from the side of the ATS connection handler (aggregate results).

Reply:
> Or you can add an extra struct/class to handle everything separately - with thread safe access from the side of the PTS connection handlers (update values) and from the side of the ATS connection handler (aggregate results).

To clarify, you mean a new inner class belongs to pager_server, and a member variable of it. Is it correct?